### ISSUE TITLE
AUPR ranks constant and all-zeros detectors above a random one, and the docstring example is wrong

### ISSUE THREAD

**Issue by mateenali66** [open-edge-platform/anomalib#3781]
`AUPR.compute()` integrates the precision-recall curve with `auc(rec, prec, reorder=True)`. A PR curve is not a function of recall, so sorting the points by recall and trapezoid-integrating does not give the area under it. The effect is large, and it is already visible in the module's own documentation.

## The module's own example is wrong

From the docstring at the top of `src/anomalib/metrics/aupr.py`:

```
>>> aupr = AUPR(fields=["pred_score", "gt_label"])
>>> aupr(batch)
tensor(0.8750)
```

with `pred_score=[0.1, 0.2, 0.8, 0.9]` and `gt_label=[0, 0, 1, 1]`. Every anomaly is ranked above every normal sample, so that is a perfect detector and the average precision is 1.0. The documented 0.8750 reproduces exactly.

## Measured

n=2000, prevalence 0.05, so the chance level is 0.05. torch 2.14.0, torchmetrics 1.9.0, current `main`.

| detector | AUPR as shipped | average precision |
|---|---|---|
| perfect, distinct scores | 0.9952 | 1.0000 |
| random, distinct scores | 0.0474 | 0.0433 |
| constant 0.5 | 0.5250 | 0.0500 |
| all-zeros | 0.5250 | 0.0500 |
| inverted, worst possible | 0.5250 | 0.0500 |

A detector that never fires, a detector that returns a single number, and the worst detector that can be constructed all score 0.5250. An honest random detector scores 0.0474. On this metric the three useless ones outrank the merely uninformative one by a factor of eleven, and they land near the middle of the scale, where a reader reasonably takes 0.5 to mean "no better than chance".

The mechanism is that a degenerate score vector has few distinct values, so the curve has few points, and traversing those points in recall order integrates a shape that has nothing to do with detection quality.

## Reproducing

```python
import torch
from torchmetrics.classification import BinaryPrecisionRecallCurve
from torchmetrics.utilities.compute import auc
from torchmetrics.functional.classification import binary_average_precision

class AUPRAsShipped(BinaryPrecisionRecallCurve):
    def compute(self):
        prec, rec, _ = super().compute()
        return auc(rec, prec, reorder=True)

g = torch.Generator().manual_seed(0)
n, k = 2000, 100
target = torch.zeros(n, dtype=torch.long)
target[torch.randperm(n, generator=g)[:k]] = 1

for name, preds in [
    ("constant", torch.full((n,), 0.5)),
    ("all-zeros", torch.zeros(n)),
    ("inverted", 1.0 - target.float()),
    ("random", torch.rand(n, generator=g)),
]:
    m = AUPRAsShipped()
    m.update(preds, target)
    print(name, float(m.compute()), float(binary_average_precision(preds, target)))
```

## Why this one and why now

The same class of defect one file over, `aupro.py` normalising by `fpr[-1]` rather than by the integration limit, was reported in https://github.com/open-edge-platform/anomalib/issues/1651 and https://github.com/open-edge-platform/anomalib/issues/2442 and fixed by https://github.com/open-edge-platform/anomalib/pull/3115. AUPRO also has `tests/unit/metrics/aupro/test_aupro.py` with a reference implementation. There is no `test_aupr.py` anywhere in the repository.

## Suggested fix

Return step-wise average precision, which is the area under the PR curve and is what `torchmetrics.functional.classification.binary_average_precision` already computes. Add `test_aupr.py` asserting that a perfect ranking returns 1.0 and that constant, all-zeros and inverted detectors land at the prevalence floor rather than at 0.5250.

That changes reported numbers on a published metric, so whether it lands as a straight fix or behind a flag is your call. I can open the PR either way.

I have not used the bug template because its Dataset, Model, Configuration YAML and Logs fields do not apply to a metric computation. https://github.com/open-edge-platform/anomalib/issues/3718 is filed the same way.

**Comment by rajeshgangireddy** [comment]
Hi @mateenali66 
Thanks for the detailed analysis and bug report. I agree and i was able to reproduce it on our side.
Would you please make a PR with your suggestions? 

Rather than calling the functional `binary_average_precision(preds, target)` (which re-derives the PR curve from scratch, duplicating the curve computation _AUPR already does for plotting), why not  inline the same formula directly in _AUPR.compute(), reusing the existing _compute() helper:


```
def compute(self) -> torch.Tensor:
    prec, rec = self._compute()
    prec = torch.where(torch.isnan(prec), torch.zeros_like(prec), prec)
    rec = torch.where(torch.isnan(rec), torch.zeros_like(rec), rec)
    return -torch.sum((rec[1:] - rec[:-1]) * prec[:-1])

```

What do you think of this? 



> That changes reported numbers on a published metric, so whether it lands as a straight fix or behind a flag is your call. I can open the PR either way.

We will include this in the changelog as well as release notes, so, no flag needed.
